src/main/java/org/apache/spark/shuffle/ucx/UcxNode.java

```diff
@@ -0,0 +1,198 @@
+
+  public UcxNode(UcxShuffleConf conf, boolean isDriver) {
+    this.conf = conf;
+    UcpParams params = new UcpParams().requestTagFeature()
+      .requestRmaFeature().requestWakeupFeature()
+      .setMtWorkersShared(true);
+    context = new UcpContext(params);
+    memoryPool = new MemoryPool(context, conf);
+    globalWorker = context.newWorker(workerParams);
+    InetSocketAddress driverAddress = new InetSocketAddress(conf.driverHost(), conf.driverPort());
+
+    if (isDriver) {
+      startDriver(driverAddress);
+    } else {
+      startExecutor(driverAddress);
+    }
+
+    // Global listener thread, that keeps lazy progress for connection establishment
+    listenerProgressThread = new UcxListenerThread(this, isDriver);
+    listenerProgressThread.start();
+
+    if (!isDriver) {
+      memoryPool.preAlocate();
+    }
+    threadLocalWorker = ThreadLocal.withInitial(() -> {
+      UcpWorker localWorker = context.newWorker(workerParams);
+      UcxWorkerWrapper result = new UcxWorkerWrapper(localWorker,
+        conf, allocatedWorkers.size());
+      if (result.id() > conf.coresPerProcess()) {
+        logger.warn("Thread: {} - creates new worker {} > numCores",
+          Thread.currentThread().getId(), result.id());
+      }
+      allocatedWorkers.add(result);
+      return result;
+    });
+  }
+
+  private void startDriver(InetSocketAddress driverAddress) {
+    // 1. Start listener on a driver and accept RPC messages from executors with their
+    // worker addresses
+    UcpListenerParams listenerParams = new UcpListenerParams().setSockAddr(driverAddress);
+    listener = globalWorker.newListener(listenerParams);
+    logger.info("Started UcxNode on {}", driverAddress);
+  }
+
+  private void startExecutor(InetSocketAddress driverAddress) {
+    // 1. Executor: connect to driver using sockaddr
+    // and send it's worker address followed by BlockManagerID.
+    globalDriverEndpoint = globalWorker.newEndpoint(
+      new UcpEndpointParams().setSocketAddress(driverAddress).setPeerErrorHadnlingMode()
+    );
+    BlockManagerId blockManagerId = SparkEnv.get().blockManager().blockManagerId();
+    ByteBuffer workerAddresss = globalWorker.getAddress();
+
+    RegisteredMemory metadataMemory = memoryPool.get(conf.metadataRPCBufferSize());
+    ByteBuffer metadataBuffer = metadataMemory.getBuffer();
+    metadataBuffer.order(ByteOrder.nativeOrder());
+    metadataBuffer.putInt(workerAddresss.capacity());
+    metadataBuffer.put(workerAddresss);
+    try {
+      SerializableBlockManagerID.serializeBlockManagerID(blockManagerId, metadataBuffer);
+    } catch (IOException e) {
+      String errorMsg = String.format("Failed to serialize %s: %s", blockManagerId,
+        e.getMessage());
+      throw new UcxException(errorMsg);
+    }
+    metadataBuffer.clear();
+    // TODO: send using stream API when it would be available in jucx.
+    globalDriverEndpoint.sendTaggedNonBlocking(metadataBuffer, new UcxCallback() {
```

Comment:
the send operation must be blocking because there is no extra reference for metadataBuffer